phpBB/phpbb/filesystem/filesystem.php

```diff
@@ -0,0 +1,904 @@
+	 */
+	protected function phpbb_is_writable($file)
+	{
+		if (file_exists($file))
+		{
+			// Canonicalise path to absolute path
+			$file = $this->realpath($file);
+
+			if (is_dir($file))
+			{
+				// Test directory by creating a file inside the directory
+				$result = @tempnam($file, 'i_w');
+
+				if (is_string($result) && file_exists($result))
+				{
+					unlink($result);
+
+					// Ensure the file is actually in the directory (returned realpathed)
+					return (strpos($result, $file) === 0) ? true : false;
+				}
+			}
+			else
+			{
+				$handle = @fopen($file, 'w');
+
+				if (is_resource($handle))
+				{
+					fclose($handle);
+					return true;
+				}
+			}
+		}
+		else
+		{
+			// file does not exist test if we can write to the directory
+			$dir = dirname($file);
+
+			if (file_exists($dir) && is_dir($dir) && $this->phpbb_is_writable($dir))
+			{
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Try to resolve real path when PHP's realpath failes to do so
+	 *
+	 * @param string	$path
+	 * @return bool|string
+	 */
+	protected function phpbb_own_realpath($path)
+	{
+		// Replace all directory separators with '/'
+		$path = str_replace(DIRECTORY_SEPARATOR, '/', $path);
+
+		$is_absolute_path = false;
+		$path_prefix = '';
+
+		if ($this->is_absolute_path($path))
+		{
+			$is_absolute_path = true;
+		}
+		else
+		{
+			// Resolve working directory and store it
+			if (is_null($this->working_directory))
+			{
+				if (function_exists('getcwd'))
+				{
+					$this->working_directory = str_replace(DIRECTORY_SEPARATOR, '/', getcwd());
+				}
+
+				//
+				// From this point on we really just guessing
+				// If chdir were called we screwed
+				//
+				else if (function_exists('debug_backtrace'))
+				{
+					$call_stack = debug_backtrace(0);
+					$this->working_directory = str_replace(DIRECTORY_SEPARATOR, '/', dirname($call_stack[sizeof($call_stack) - 1]['file']));
+				}
+				else
+				{
+					//
+					// Assuming that the working directory is phpBB root
+					// we could use this as a fallback, when phpBB will use controllers
+					// everywhere this will be a safe assumption
+					//
+					//$dir_parts = explode(DIRECTORY_SEPARATOR, __DIR__);
+					//$namespace_parts = explode('\\', trim(__NAMESPACE__, '\\'));
+
+					//$namespace_part_count = sizeof($namespace_parts);
+
+					// Check if we still loading from root
+					//if (array_slice($dir_parts, -$namespace_part_count) === $namespace_parts)
+					//{
+					//	$this->working_directory = implode('/', array_slice($dir_parts, 0, -$namespace_part_count));
+					//}
+					//else
+					//{
+					//	$this->working_directory = false;
+					//}
+
+					$this->working_directory = false;
+				}
+			}
+
+			if ($this->working_directory !== false)
+			{
+				$is_absolute_path = true;
+				$path = $this->working_directory . '/' . $path;
+			}
+		}
+
+		if ($is_absolute_path)
+		{
+			if (strpos(strtolower(PHP_OS), 'win') !== false)
```

Comment:
Someone knows what is the best way to replace this? defined('PHP_WINDOWS_VERSION_MAJOR')?